Here are some of the failed tests for example. Looks quite strange. I've upgraded to python 3.14 and latest jinja2==3.1.6

```
  tojson ensure_ascii=true
    Python script failed with exit code: 1
    Output: Traceback (most recent call last):
  File "<string>", line 14, in <module>
(REDACTED)
    raise rewrite_traceback_stack(source=source)
  File "<template>", line 1, in top-level template code
TypeError: do_tojson() got an unexpected keyword argument 'ensure_ascii'

  ASSERTION FAILED : python execution
  tojson ensure_ascii=true (1 of 1 assertion(s) failed)                         [FAIL]
  tojson sort_keys=true
    Python script failed with exit code: 1
    Output: Traceback (most recent call last):
  File "<string>", line 14, in <module>
(REDACTED)
    raise rewrite_traceback_stack(source=source)
  File "<template>", line 1, in top-level template code
TypeError: do_tojson() got an unexpected keyword argument 'sort_keys'

  ASSERTION FAILED : python execution
  tojson separators=(',',':') (1 of 1 assertion(s) failed)                      [FAIL]
  tojson separators=(',',': ') indent=2
    Python script failed with exit code: 1
    Output: Traceback (most recent call last):
  File "<string>", line 14, in <module>
(REDACTED)
    raise rewrite_traceback_stack(source=source)
  File "<template>", line 1, in top-level template code
TypeError: do_tojson() got an unexpected keyword argument 'separators'
```

--- a/tests/test-jinja.cpp
+++ b/tests/test-jinja.cpp
@@ -1115,16 +1115,23 @@ static void test_template_cpp(testing & t, const std::string & name, const std::
 
 static std::string py_script = R"(
 import jinja2
+import jinja2.ext as jinja2_ext
 import json
 import sys
+from datetime import datetime
+from jinja2.sandbox import ImmutableSandboxedEnvironment
 
 tmpl = json.loads(sys.argv[1])
 vars_json = json.loads(sys.argv[2])
 
-env = jinja2.Environment(
+env = ImmutableSandboxedEnvironment(
     trim_blocks=True,
     lstrip_blocks=True,
+    extensions=[jinja2_ext.loopcontrols],
 )
+environment.filters["tojson"] = lambda x, ensure_ascii=False, indent=None, separators=None, sort_keys=False: json.dumps(x, ensure_ascii=ensure_ascii, indent=indent, separators=separators, sort_keys=sort_keys)
+environment.globals["strftime_now"] = lambda format: datetime.now().strftime(format)
+environment.globals["raise_exception"] = lambda message: raise jinja2.exceptions.TemplateError(message)
 template = env.from_string(tmpl)
 result = template.render(**vars_json)
 print(result, end='')

--- a/tests/test-jinja.cpp
+++ b/tests/test-jinja.cpp
@@ -1113,6 +1113,7 @@ static void test_template_cpp(testing & t, const std::string & name, const std::
     });
 }
 
+// keep this in-sync with https://github.com/huggingface/transformers/blob/main/src/transformers/utils/chat_template_utils.py
 static std::string py_script = R"(
 import jinja2
 import jinja2.ext as jinja2_ext
@@ -1129,9 +1130,14 @@ env = ImmutableSandboxedEnvironment(
     lstrip_blocks=True,
     extensions=[jinja2_ext.loopcontrols],
 )
-environment.filters["tojson"] = lambda x, ensure_ascii=False, indent=None, separators=None, sort_keys=False: json.dumps(x, ensure_ascii=ensure_ascii, indent=indent, separators=separators, sort_keys=sort_keys)
-environment.globals["strftime_now"] = lambda format: datetime.now().strftime(format)
-environment.globals["raise_exception"] = lambda message: raise jinja2.exceptions.TemplateError(message)
+
+def raise_exception(message):
+    raise jinja2.exceptions.TemplateError(message)
+
+env.filters["tojson"] = lambda x, ensure_ascii=False, indent=None, separators=None, sort_keys=False: json.dumps(x, ensure_ascii=ensure_ascii, indent=indent, separators=separators, sort_keys=sort_keys)
+env.globals["strftime_now"] = lambda format: datetime.now().strftime(format)
+env.globals["raise_exception"] = raise_exception
+
 template = env.from_string(tmpl)
 result = template.render(**vars_json)
 print(result, end='')

--- a/tests/test-jinja.cpp
+++ b/tests/test-jinja.cpp
@@ -1114,18 +1114,19 @@ static void test_template_cpp(testing & t, const std::string & name, const std::
 }
 
 // keep this in-sync with https://github.com/huggingface/transformers/blob/main/src/transformers/utils/chat_template_utils.py
+// note: we use SandboxedEnvironment instead of ImmutableSandboxedEnvironment to allow usage of in-place array methods like append() and pop()
 static std::string py_script = R"(
 import jinja2
 import jinja2.ext as jinja2_ext
 import json
 import sys
 from datetime import datetime
-from jinja2.sandbox import ImmutableSandboxedEnvironment
+from jinja2.sandbox import SandboxedEnvironment
 
 tmpl = json.loads(sys.argv[1])
 vars_json = json.loads(sys.argv[2])
 
-env = ImmutableSandboxedEnvironment(
+env = SandboxedEnvironment(
     trim_blocks=True,
     lstrip_blocks=True,
     extensions=[jinja2_ext.loopcontrols],

